### DISCUSSION TITLE
TypeError: Cannot read properties of null (reading 'cssRules')

### DISCUSSION THREAD

**Discussion by tmcconechy** [percy/cli#1212] [Support]
## The problem

Upgraded from 1.18.0 to 1.20.1 and now get this error on most tests:

```
[percy] Could not take DOM snapshot "ids-data-grid-new-light"
[percy] Error: Evaluation failed: TypeError: Cannot read properties of null (reading 'cssRules')
    at styleSheetsMatch (pptr://__puppeteer_evaluation_script__:177:50)
    at serializeCSSOM (pptr://__puppeteer_evaluation_script__:210:15)
    at serializeElements (pptr://__puppeteer_evaluation_script__:465:9)
    at Object.serializeDOM (pptr://__puppeteer_evaluation_script__:504:7)
    at pptr://__puppeteer_evaluation_script__:3:23
```    

## Environment

- Node version:
- `@percy/cli` version: 1.20.1
- Version of Percy SDK you’re using:
- If needed, a build or snapshot ID: https://github.com/infor-design/enterprise-wc/actions/runs/4315837554/jobs/7530775761#step:10:267
- OS version: Unix/Mac
- Type of shell command-line [interface]: bash

## Details

Example test https://github.com/infor-design/enterprise-wc/blob/main/test/ids-data-grid/ids-data-grid-percy-test.ts

## Debug logs

https://github.com/infor-design/enterprise-wc/actions/runs/4315837554/jobs/7530775761#step:10:267

## Code to reproduce issue

If needed:
- clone https://github.com/infor-design/enterprise-wc
- change `"@percy/cli": "1.18.0",` to `"@percy/cli": "1.20.1",` in package.json
- `npm i && npm run test`

**Comment by tmcconechy** [comment]
@itsjwala this fixed the cssRules error but i still cant use the latest version it gives some unknown errors now not showing up in the log https://github.com/infor-design/enterprise-wc/actions/runs/4365577412/jobs/7637547882#step:10:29623

Just mentioning it in case you have an idea. If not i can try to get verbose logs turned on (is that possible I forget) when i have more time and make a new issue.

**Comment by samarsault** [comment]
@tmcconechy I see the visual changes are happening due to javascript being enabled. Is there any reason for enabling it? (If its Shadow DOM, then we support it since 1.19.2+) We noticed that the visual errors are a result of javascript side effects. Can you try disabling it and running the builds?

**Comment by tmcconechy** [comment]
@samarsault i disabled it on https://github.com/infor-design/enterprise-wc/pull/1165 but it a) still has a lot of errors b) the images dont look correct anymore.

We do use shadowRoot (web components). But not sure i recall why we enabled it. I think support suggested it. Still this only is a problem moving up from 1.18.0

https://github.com/infor-design/enterprise-wc/actions/runs/4375912710

**Comment by samarsault** [comment]
@tmcconechy we see some unusual diffs in your build and will take a look. Can you try using the latest version setting `enableJavascript:true` and `disableShadowDOM:true`?

**Comment by tmcconechy** [comment]
@samarsault still doing some weird things https://github.com/infor-design/enterprise-wc/actions/runs/4397781859/jobs/7716169624 https://github.com/infor-design/enterprise-wc/pull/1165

I would think we want JS and shadowRoot since these are JS components and they use web components/shadowRoot

**Comment by tmcconechy** [comment]
@samarsault should i be using the support at browser stack instead? My concern is we can no longer update and use percy so might need to look at other options before renewing.

**Comment by samarsault** [comment]
@tmcconechy For the enableJavascript: false case, I see the following classes of visual changes in your build

- Date & Time changes in tables : These are expected because javascript is no longer running in the rendering environment
- Width changes: This can happen If you handle responsive changes in your javascript. See [this doc](https://docs.percy.io/docs/capturing-responsive-dom-snapshots#examples) for reference.

Also, is the cypress test failure happening because of percy? If yes, what is the error message?
If you still face any issue, please reach out through support. It would help us in prioritizng this.

Edit:
For the `enableJavascript: true` case, we've identified an issue with shadow dom + enableJavascript: true configuration. Fixing this in #1210

**Comment by tmcconechy** [comment]
ok thanks @samarsault cant seem to get it to work with any configuration i can come up with. We use JS and and its all web components so Js off is not an option (will show nothing). Shadow root is also needed, without it will show nothing on the page. This is the last build https://github.com/infor-design/enterprise-wc/pull/1165 every time it just comes up with errors on each test i cant figure out. So will log a support ticket for now. Cheers

**Comment by samarsault** [comment]
@tmcconechy I have discussed this with my team and we're checking this issue on priority.

**Reply by tmcconechy** [reply]
Ok thanks. Let me know if i can provide anything more to you...

**Reply by samarsault** [reply]
How do I run percy builds locally in your project? 

I tried `npx percy exec -- npm run test:ci` but it `gives a window not defined error - consider using jsdom test environment`. Same goes for `npm run test:percy:local`. Setting the jsdom variable introduces a `page not defined` error, which I'm guessing is because cypress was not initialized. I look at your github action but could not find any relevant step about this.

We've identified a fix and I want to test it before sharing

**Reply by samarsault** [reply]
@tmcconechy ^

**Comment by tmcconechy** [comment]
@samarsault you could run it one of two ways

```
npx percy exec -- npx jest -- percy
```

or just like this all the tests for percy are labeled with percy in the name
```
 npx jest -- percy
```

We dont use cypress at the moment its jest/jest puppetteer. I've never seen that error before. Did you do `npm i`? Are you on mac or windows?

**Reply by samarsault** [reply]
@tmcconechy I'm on mac. The following run is on the `Sync local` commit of your repo with a modification of testEnvironment: "jsdom".

<img width="1129" alt="Screenshot 2023-03-17 at 9 43 07 PM" src="https://user-images.githubusercontent.com/2987087/225959685-9a3fd5c4-654b-4018-8c25-72e04cc47d1b.png">

Without the modification of `testEnvironment: 'jsdom'` it gives the following error

<img width="1352" alt="Screenshot 2023-03-17 at 9 46 10 PM" src="https://user-images.githubusercontent.com/2987087/225960378-6db34d10-7f7c-480e-82fb-38eec60b7240.png">

I tried `npx percy exec -- npx jest -- percy` and `npx percy exec -- npx jest -- percy` but the results were the same.

**Comment by tmcconechy** [comment]
Hmm when i try those it works ok. Im not 100% sure. Maybe try `rm -rf node_modules && npm i` again and make sure it didnt give any errors/things missing? It should be using puppetteer not jsdom for these "e2e" tests. We treat percy tests as e2e tests.

Failing that maybe if you want i can test the fix? Or you could try to test it by modifying the PR?

**Reply by samarsault** [reply]
@tmcconechy Can you try [this patch](https://github.com/percy/cli/pull/1210) with `enableJavascript: true` and `disableShadowDOM: true`. I tested it with a few snapshots but am unable to do on the whole build. Note that the `disableShadowDOM` option is only to let Percy know not to serialize the shadow dom since its being done by JS

**Reply by tmcconechy** [reply]
@samarsault would it be possible to put this in a release? That would be the easiest way to try it on the CI server. I think it seems logical.

I tried to build percy on that branch but not sure how to consume it? If i do `yarn build` how can i copy that to my local branch and use it?

**Reply by itsjwala** [reply]
Hey @tmcconechy 

* Just check the developing section in readme.
* Secondly you could also check `test.yml` github actions in `percy-cypress` repo

**Reply by tmcconechy** [reply]
Hmm not sure what im looking for in the developing section? Can you point me at the section? I have it built. but not sure what i just built. I.E. how would i copy the files to node_modules to test.

If you are ok with the fix lets try it in a release?

**Reply by itsjwala** [reply]
* post cloning the `percy/cli` repo, run following steps https://github.com/percy/percy-cypress/blob/master/.github/workflows/test.yml#L35-L44  (TLDR;  `yarn` &  `yarn build`) 
* We do `yarn global:link`  - [ref](https://github.com/percy/cli#developing) for making this clone repo available globally in your local system.  (This needs to be done in CLI repo)
* Post this is done, just link the required package in your git repository using `yarn link @percy/dom` ( since only dom package changes are required from this https://github.com/percy/cli/pull/1210)

**Reply by tmcconechy** [reply]
OK i was able to test this out i think. But i still get a lot of errors.

```
Test Suites: 6 failed, 4 skipped, 66 passed, 72 of 76 total
Tests:       37 failed, 13 skipped, 227 passed, 277 total
```

And not sure if this is anything:
```
ts-jest[versions] (WARN) Version 5.0.2 of typescript installed has not been tested with ts-jest. If you're experiencing issues, consider using a supported version (>=4.3.0 <5.0.0-0). Please do not report issues in ts-jest if you are using unsupported versions.
```

If you want to just try releasing i can try on the CI server which is the usual way we do it. Im not sure how to verify the link did work so not sure if I'm on the right version. (Usually use npm but i followed the steps).

BTW ran with `npx percy exec -- npx jest -- percy` after setting tokens and setup as per above. `yarn link @percy/dom` after building and linkin on branch `enable-js-shadow`

**Reply by itsjwala** [reply]
> If you want to just try releasing i can try on the CI server which is the usual way we do it.

There are some changes that we're yet to finish in that PR

> Im not sure how to verify the link did work so not sure if I'm on the right version.

Following command should list the percy packages you're using, `dom` would be linked to the global one which you cloned before.
```sh
ls -lhart node_modules/@percy/
```

btw, you could try the same steps in Github Actions as we did in `percy-cypress` repo

**Reply by itsjwala** [reply]
Also, I was checking your https://github.com/infor-design/enterprise-wc/pull/1185/files,  we've changed the param name from `disableShadowDOM` to `disableShadowDOMSerialization` to avoid confusion.

**Reply by tmcconechy** [reply]
Ok fixed the rename. I tried to run that command and i get:

```
tim.mcconechy@USROMTMCCONEC02 enterprise-wc4 % ls -lhart node_modules/@percy/
total 0
drwxr-xr-x@   7 tim.mcconechy  staff   224B Mar 10 10:59 puppeteer
drwxr-xr-x@   7 tim.mcconechy  staff   224B Mar 10 10:59 sdk-utils
drwxr-xr-x@   6 tim.mcconechy  staff   192B Mar 21 09:42 cli-snapshot
drwxr-xr-x@   6 tim.mcconechy  staff   192B Mar 21 09:42 cli-app
drwxr-xr-x@   6 tim.mcconechy  staff   192B Mar 21 09:42 cli-upload
drwxr-xr-x@   6 tim.mcconechy  staff   192B Mar 21 09:42 cli-exec
drwxr-xr-x@   6 tim.mcconechy  staff   192B Mar 21 09:42 cli-build
drwxr-xr-x@   6 tim.mcconechy  staff   192B Mar 21 09:42 cli-config
drwxr-xr-x@   7 tim.mcconechy  staff   224B Mar 21 09:42 logger
drwxr-xr-x@   7 tim.mcconechy  staff   224B Mar 21 09:42 env
drwxr-xr-x@   7 tim.mcconechy  staff   224B Mar 21 09:42 client
drwxr-xr-x@   8 tim.mcconechy  staff   256B Mar 21 09:42 cli-command
drwxr-xr-x@   8 tim.mcconechy  staff   256B Mar 21 09:42 config
drwxr-xr-x@ 869 tim.mcconechy  staff    27K Mar 21 09:42 ..
lrwxr-xr-x@   1 tim.mcconechy  staff    40B Mar 21 09:43 dom -> ../../../../.config/yarn/link/@percy/dom
drwxr-xr-x@  18 tim.mcconechy  staff   576B Mar 21 09:43 .
drwxr-xr-x@   8 tim.mcconechy  staff   256B Mar 21 09:43 cli
drwxr-xr-x@  11 tim.mcconechy  staff   352B Mar 21 09:45 core
```

Not sure what it means but looks like the link... I cant try gihub actions as far as i know until this is released. So i guess i'll wait.

**Reply by itsjwala** [reply]
* This PR will take some time to get merged.
* For now, I just pushed a prerelease [1.22.0-alpha.0](https://www.npmjs.com/package/@percy/cli/v/1.22.0-alpha.0) you could give it a try!

**Reply by itsjwala** [reply]
@tmcconechy I tried to run your project locally but it hangs does this require some setup to be done?


```bash
➜  enterprise-wc git:(main) cat package.json | grep percy
    "test:ci": "npm run test:ui && npm run test:percy",
    "test:percy": "npx jest --verbose --runInBand --detectOpenHandles --logHeapUsage --forceExit -- percy",
    "test:percy:local": "npx percy exec -- npm run test:percy",
    "@percy/cli": "1.18.0",
    "@percy/puppeteer": "2.0.2",

➜  enterprise-wc git:(main) npm run test:percy:local

> ids-enterprise-wc@1.0.0-beta.8 test:percy:local
> npx percy exec -- npm run test:percy


[percy] A new version of @percy/cli is available! 1.18.0 -> 1.21.0

[percy] Percy has started!
[percy] Running "npm run test:percy"

> ids-enterprise-wc@1.0.0-beta.8 test:percy
> npx jest --verbose --runInBand --detectOpenHandles --logHeapUsage --forceExit -- percy

Dev server listening on port 4444

 RUNS  test/ids-data-grid/ids-data-grid-percy-test.ts

```

**Reply by tmcconechy** [reply]
Ok i can wait for when its completed. No problem I will just stay on the old version.

Maybe when running this you need to do a build first to build the site that it runs... Try this sequence:

```
git checkout new-deps-updates
git pull
npm prune && npm i // on node 18 nvm use to use node 18
npm run build
npx percy exec -- npx jest -- percy // just percy tests nothing else
```

**Reply by itsjwala** [reply]
* @tmcconechy could you try the prerelease till then(so we know the patch works) 
* I'm trying your steps locally till then, thanks!

**Reply by tmcconechy** [reply]
ok its running on the pre-release https://github.com/infor-design/enterprise-wc/pull/1185 i will check back a bit later as it takes about 40 mins.

**Reply by tmcconechy** [reply]
Looks like its doing the same thing https://github.com/infor-design/enterprise-wc/actions/runs/4491972641/jobs/7901194439#step:10:300 ?

Should i be turning on some logging somehow? Would that help?

**Reply by tmcconechy** [reply]
I also ran it isolated here https://github.com/infor-design/enterprise-wc/actions/runs/4493463030/jobs/7904705397 same issue.
Just to see if its other dependencies or anything

**Reply by itsjwala** [reply]
@tmcconechy 

I tried running DOM snapshot on an isolated snapshot `ids-masthead` and I could see this error is raised, it seems your web component framework is detecting a change in the DOM, and throwing an error( though this doesn't interfere with Percy Snapshotting and visual testing) 

```js
ids-menu-button.ts?b6fe:79 Uncaught TypeError: Cannot set properties of null (setting 'disabled')
    at set disabled [as disabled] (ids-menu-button.ts?b6fe:79:17)
    at HTMLElement.attributeChangedCallback (ids-element.ts?5d20:92:41)
    at HTMLElement.onHiddenChange (ids-button.ts?aea2:202:12)
    at HTMLElement.attributeChangedCallback (ids-element.ts?5d20:73:14)
    at set colorVariant [as colorVariant] (ids-color-variant-mixin.ts?aeec:67:14)
    at HTMLElement.attributeChangedCallback (ids-element.ts?5d20:92:41)
    at set mode [as mode] (ids-theme-mixin.ts?2db2:63:10)
    at HTMLElement.attributeChangedCallback (ids-element.ts?5d20:92:41)
    at set menu [as menu] (ids-menu-button.ts?b6fe:160:10)
    at HTMLElement.attributeChangedCallback (ids-element.ts?5d20:92:41)
```

tested with starting local Percy server

```
npx percy exec:start --debug
```

and doing steps as follow in devtools

<img width="1084" alt="image" src="https://user-images.githubusercontent.com/24763310/227192895-d47a5708-4515-48b4-9254-0f8fba47ae15.png">
-----

since in the new release we're doing serialization of Shadow DOM - it requires `enableJavascript: false`, even with `enableJavascript: true` `disableShadowDOM: true` the framework is causing side effects on the DOM.

with the same steps as above ^ image, you could try printing 
`console.log(PercyDOM.serialize({enableJavaScript:true, disableShadowDOM: true}).html)`   and you'll observe double icons and some other side effects. 

-----

currently to even render all the snapshots visually properly setting `enableJavaScript: false` in your `.percy.json` is the best way to do it. 

Note: we're serializing shadow roots and sending in infrastructure to render without any flakes, so you don't need javascript enabled here, also no `disableShadowDOM` setting is required either from `v1.19.2+` onwards - [documentation](https://docs.percy.io/docs/shadow-dom)

--- 

I tried on my project https://percy.io/9560f98d/infor-repro **( I'd to disable CSP for fonts and CSS being set in [here](https://github.com/infor-design/enterprise-wc/blob/5640bff0930ddffa1c09965a52a5eafb1a8cd6a3/scripts/webpack-html-templates.js) and increase Jest timeout to `25s` to run it locally)**

* As you could see build 4 - 6 runs on `v1.18.0` and it flakes in some snapshots (due to JS causing side effects) `enableJavaScript: true`

* build ran using `enableJavaScript: false` on `v1.21.0` from 8, 10 have lesser flakes.
  * The diffs that you see are one-time.
  * Like [this](https://percy.io/9560f98d/infor-repro/builds/26117588/changed/1455651694?browser=chrome&browser_ids=34&subcategories=approved&viewLayout=overlay&viewMode=new&width=1280&widths=375%2C1280) diff for example(and similar ones), can be solved by resizing the browser to `1280` width and then calling `percySnapshot`.

**Reply by tmcconechy** [reply]
Ok it doesnt seem to be our code. You can also test https://main.wc.design.infor.com/ids-masthead/example.html 
or if you run `npm run start` and go to http://localhost:4300/ids-masthead/example.html it does work.

Again this all started after 1.18 of percy. Seems like it should be interrupting the code so much to me. It should just scrape the page and take a screen shot whatever we show on the page?

From that last comment is there something i should do?

**Reply by itsjwala** [reply]
> Ok it doesnt seem to be our code. You can also test https://main.wc.design.infor.com/ids-masthead/example.html
or if you run npm run start and go to http://localhost:4300/ids-masthead/example.html it does work.

Please try with `npx percy exec:start --debug` (I'd added steps above ^) 

> It should just scrape the page and take a screen shot whatever we show on the page?

Please check this doc https://docs.percy.io/docs/debugging-sdks#how-percy-works for better understanding. 

* As I mentioned in the last part ^, from `1.19.2+` onwards we're supporting Shadow DOM serialization, so only having `enableJavaScript: false` and removing `disableShadowDOM` should work, locally I didn't have an issue in creating builds. 
* `enableJavaScript: true` will cause issues with the newer releases for you.

---

btw, I'd suggest looking at this build - https://percy.io/9560f98d/infor-repro/builds/26117588 ( taken with js disable and `v1.21.0`)
It has lot many meaningful diffs which in `1.18.0` were not caught.

**Reply by tmcconechy** [reply]
Ok trying this https://github.com/infor-design/enterprise-wc/pull/1201/files#diff-0015438c1c3b14d8068b71e74ac15e9943d0a23b1cb548c25705962cb17231d6L9 

If i try your steps i get:

```
tim.mcconechy@USROMTMCCONEC02 enterprise-wc4 % npx percy exec:start --debug

[percy] Heads up! The current version of @percy/cli is more than 10 releases behind! 1.22.0-alpha.0 -> 1.21.0

[percy:config] Found config file: .percy.json (288ms)
[percy:config] Using config:
{
  version: 2,
  snapshot: {
    widths: [
      375,
      1280
    ],
    minHeight: 1024,
    enableJavaScript: false
  }
} (6ms)
[percy:core:install] Successfully downloaded Chromium 929475 (4945ms)
[percy:core:browser] Launching browser (1ms)
[percy:core:browser] Browser connected [6629]: HeadlessChrome/96.0.4664.0 (1394ms)
[percy:core] Percy has started! (2ms)
```

Then it just hangs...But from the error message it seems wierd that your code injects into our code and causes issues when it works normally.

Im ok with changing any config. But this was all working beautifully before 1.18. Seems like enableJavascript would be needed but not sure what that means on your end. Just that our code uses JS extensively and uses shadowDoms due to it being a "web component".

Lets see what https://github.com/infor-design/enterprise-wc/pull/1201/files#diff-0015438c1c3b14d8068b71e74ac15e9943d0a23b1cb548c25705962cb17231d6L9  does

**Reply by itsjwala** [reply]
> Then it just hangs...But from the error message it seems wierd that your code injects into our code and causes issues when it works normally.

oops I forgot to write the steps down explicitly, check the screenshot ^

```js
eval(await (await fetch('http://localhost:5338/percy/dom.js')).text())
console.log(PercyDOM.serialize().html)
```


> Im ok with changing any config. But this was all working beautifully before 1.18. Seems like enableJavascript would be needed but not sure what that means on your end. Just that our code uses JS extensively and uses shadowDoms due to it being a "web component".

for your context https://github.com/percy/cli/pull/1165

**Reply by tmcconechy** [reply]
Hmm yeah it still doesnt work with enableJS off and disableShadow gone. https://github.com/infor-design/enterprise-wc/pull/1201/files#diff-0015438c1c3b14d8068b71e74ac15e9943d0a23b1cb548c25705962cb17231d6L9

I'm not sure what i can do. But I would expect we could make it work as it did before or at least capture when components that use shadowDOM still. https://github.com/percy/cli/pull/1165 sounds great but it seems to have broke our use of it and our usage is using the newer Web Component standard which is growing in adoption day by day.

**Reply by itsjwala** [reply]
* The errors as you can see aren't coming from Percy ^ it's coming from internal implementation of the framework ( note this will occur now since we're serializing shadow dom which will cause the connected callbacks to be triggered and your framework seems to be reacting based on this - you need to handle this somehow in your framework)
* Speaking of Diffs most of the diff could be resolved by resizing the browser to `1280` and `375` before calling `percySnapshot`, Please give it a try - [ref](https://docs.percy.io/docs/capturing-responsive-dom-snapshots)!
* I don't see it being broken though from the Percy build which is generated ^ - most of the diffs are genuine correct diffs.

**Reply by tmcconechy** [reply]
Ok i sort of get it but still cant figure out how to debug this the way your doing percy.Serialize. I guess it does something with that page (like takes the html and copies it around?). Its possible some components need some work there on different ways of initializing it. (i.e document.createElement vs document.append()?). Also this could be a time consuming change but probably for the better...

Can you explain how to get from running `npx percy exec:start --debug` to the browser? as if i run that command and go to http://localhost:4300/ids-masthead it isnt working. So maybe help me with the steps to see it on masthead as its a bit confusing to follow.

Concerning the resize what about this: https://github.com/infor-design/enterprise-wc/blob/main/.percy.json#L4-L7 i thought this would just do 2 screen shots in those sizes? Should we remove that and do this now: https://docs.percy.io/docs/capturing-responsive-dom-snapshots

**Reply by tmcconechy** [reply]
And what about CSP? https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP since we use that. i think the eval isnt allowed.

**Reply by tmcconechy** [reply]
Ok i think i got it. Steps

1. in one window run `npx percy exec:start --debug`
2. in another run `npm run start`
3. go to http://localhost:4300/ids-masthead/example.html
4. type this in the console

```
eval(await (await fetch('http://localhost:5338/percy/dom.js')).text())
console.log(PercyDOM.serialize().html)
```

In the case of masthead i see we need an extra guard but in one case its really wierd as its disconnected children from its parent element ect.. (datagrid). So really not seeing any easy fix there.

What exactly is this serialize going to do, i read the PR but didnt get it? Is it like expecting server side rendering to work? Like say i have an element like `<ids-tag></ids-tag>` only in the page. What is it doing to it?

Any way to disable it and do it as before?

If you want try the same tests on masthead you did on http://localhost:4300/ids-data-grid/example.html -> im not sure what i would fix. Looking at the stacktrace could it be walking the tree in the wrong order?

**Reply by itsjwala** [reply]
> Concerning the resize what about this: https://github.com/infor-design/enterprise-wc/blob/main/.percy.json#L4-L7 i thought this would just do 2 screen shots in those sizes? Should we remove that and do this now: https://docs.percy.io/docs/capturing-responsive-dom-snapshots

The widths that are passed either in global config ( `.percy.json` in your case) or `percySnapshot` is used for asset discovery and rendering at that width in infrastructure browsers. but the dom snapshot which is taken is in the test browser, so you'll need to resize your browser before taking calling `percySnapshot`  

[capturing responsive dom snapshots](https://docs.percy.io/docs/capturing-responsive-dom-snapshots) is used when you have say `375` and `1280` widths where in the dom state is not same for both. 

TLDR; in your case since you're using both the widths and they have different dom states, you have to follow capturing responsive dom snapshots

Please check this doc https://docs.percy.io/docs/debugging-sdks#how-percy-works for better understanding.

> And what about CSP? https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP since we use that. I think the eval isn't allowed.

yes, we require that our dom snapshot should work without any errors in your test browser, currently it requires eval, so better it to disable CSP when running Percy tests


> What exactly is this serialize going to do, i read the PR but didnt get it? Is it like expecting server side rendering to work? Like say i have an element like <ids-tag></ids-tag> only in the page. What is it doing to it?

> Any way to disable it and do it as before?

I tried it with `disableShadowDOM` but it is still showing side effects, and we're any way gonna deprecate this flag in future since shadow dom APIs would be available widely. 

so the best way forward with the current state was `enableJavascript: false` and not using `disableShadowDOM` 


> If you want try the same tests on masthead you did on http://localhost:4300/ids-data-grid/example.html -> im not sure what i would fix. Looking at the stacktrace could it be walking the tree in the wrong order?

we use DFS preorder traversal - [ref](https://github.com/percy/cli/blob/master/packages/dom/src/clone-dom.js#L14-L46) 
Ideally the order doesn't matter - what is causing an issue is the callbacks that are triggered I think(this is handled in `Lit/Stencil` components I believe since most customers are using these frameworks when using shadow dom with percy, maybe you can have a look at their implementation. )

**Reply by tmcconechy** [reply]
Ok Will review the components. See if we can fix it, i fixed one of 32 already. It might just be some are not working with `document.create` or something. But the data-grid one did not have an obvious fix. 

Last Question For now:
- i don't 100% get what your saying about the sizes? So I should remove the config https://github.com/infor-design/enterprise-wc/blob/main/.percy.json#L4-L7 and just do https://docs.percy.io/docs/capturing-responsive-dom-snapshots#playwright-or-puppeteer instead? Or leave the config and add that anyways? Does it increase the snapshot count?
- can i run the tests in a way that shows the errors in the console? seems to hide the errors unless you run the debugger.

**Reply by itsjwala** [reply]
> i don't 100% get what your saying about the sizes? So I should remove the config https://github.com/infor-design/enterprise-wc/blob/main/.percy.json#L4-L7 and just do https://docs.percy.io/docs/capturing-responsive-dom-snapshots#playwright-or-puppeteer instead? Or leave the config and add that anyways? Does it increase the snapshot count?

Please refer to this [documentation](https://docs.percy.io/docs/cli-configuration#per-snapshot-configuration) 

> widths: An array of numbers, in pixels, representing the widths you'd like to capture for each snapshot; overrides global snapshot widths.

you can keep global config, but for snapshots where you need responsive dom snapshot do something like this(this will override global config) - 

```js
  await page.setViewport({ width: 375, height: 1024 }); 
  await percySnapshot(page, 'Home Page', { width: 375 });
  await page.setViewport({ width: 1280, height: 1024 }); 
  await percySnapshot(page, 'Home Page', { width: 1280 });
```

make sure to have in your global config

```yaml
percy:
  defer-uploads: true
```

and it won't increase your screenshot count(billing perspective), since you're we're rendering same number of screenshots in infrastructure. 

> can i run the tests in a way that shows the errors in the console? seems to hide the errors unless you run the debugger.


The issue is this error doesn't occur in the `@percy/dom` package so we can't catch and log it.

**Comment by tmcconechy** [comment]
@itsjwala - I cant seem to find a way to make this work with some of our components. Take for example `ids-data-grid` instead of `ids-masthead`. You will see it errors on ids-data-grid-row because that is expecting the parent to be the data grid. You can't have a row without a data grid.

But your code is serializing this in a weird way that causes the row children to be added before the parent. Is there any way to do the old way?

According to https://github.com/percy/cli/pull/1165/files#diff-02067f8a8ffa312638a2a6525baaabf925e44fc81392b24fc9de58604100be75R27 disableShadowDom might work (didnt in my tests). But this is going away?

Other thing that is weird is a lot of text elements are duplicated in the snap shots (see some text duplicated/icons duplicated ect)...

Maybe we can have some options for serializing here? The old way did work pretty good for us

**Reply by samarsault** [reply]
Is all of this happening in the enable javascript false case?

> I cant seem to find a way to make this work with some of our components. Take for example ids-data-grid instead of ids-masthead. You will see it errors on ids-data-grid-row because that is expecting the parent to be the data grid. You can't have a row without a data grid.

Can you share the expected and received HTML structure? Also share the snapshot link/id where this is happening

> Other thing that is weird is a lot of text elements are duplicated in the snap shots (see some text duplicated/icons duplicated ect)...

Can you share the particular snapshots where this is happening?

**Reply by tmcconechy** [reply]
Here is the job with no JS https://github.com/infor-design/enterprise-wc/pull/1218 (still some errors i didnt fix / some i really cant).
But you can see some have duplicate text:

https://images.percy.io/1fedaa89ac4a6e622ec586b8cda83f59240718437bc432a4c79bb085bd8ab0d7
https://images.percy.io/a438c045d8ca0b5da18cbb50ffa4a51acaafdb8122315eca9546d9ff63106c8f

**Comment by samarsault** [comment]
@tmcconechy Thanks for your patience. We've been looking into your issues and here's what we got:

- **Duplicate issue**: I see the enterprise wc code relies heavily on `attributeChangedCallback` . When cloning nodes while serializing, this callback is triggered and causes side effects in our clone. Since this there’s no way to stifle attributeChangedCallback in the clone, we’ve added a workaround  in https://github.com/percy/cli/pull/1230 in cli v1.23.0. Please upgrade to the same
- **Serialization Errors**. They are getting triggered in attributeChangedCallback but they fail since it expects the whole structure in place. Since we build the tree step by step this can be ignored. If you are seeing webpack errors in screenshots (in a dev environment), you can use the following percy css to hide it - 
```yml
snapshot:
   percyCSS: |
      iframe#webpack-dev-server-client-overlay {
      display: none 
    }
```
- **enableJavascript: false** As Jigar had mentioned earlier ([ref](https://github.com/percy/cli/discussions/1212#discussioncomment-5418243)), you’ll have to take responsive dom snapshots. This is needed as javascript can lead to flaky builds

- **CI Errors**: There no exception being thrown or error message being displayed. We ran through individual test cases using a debugger and `percySnapshot` completes without any error. We ran the failing tests on a clean percy-puppeteer project and it worked fine. So it seems to be some coupling issue with the test suite and percy
<img width="360" alt="Screenshot 2023-04-10 at 8 46 36 PM" src="https://user-images.githubusercontent.com/2987087/230930739-c0b6f865-2953-4ca8-bcc7-f4d000f80c56.png">

**Comment by tmcconechy** [comment]
FYI after our call i let our ci run it and it did run into errors https://github.com/infor-design/enterprise-wc/actions/runs/5015136541/jobs/8990283101

I think when i ran it local it wasnt pointing to percy and just skipping tests, so I will have to look at this some more.

**Comment by samarsault** [comment]
@tmcconechy Can you refactor all your calls to `percySnapshot` to include the `disableShadowDOM: true` config? This would fallback to the old serialization algorithm. I checked with your tests and they are passing with this

You can try something like:
```javascript
function customPercySnapshot(page, name, options) {
   return percySnapshot(page, name, { disableShadowDOM: true, ...options })
}

export default customPercySnapshot
```

And then import this as `percySnapshot` everywhere

**Reply by tmcconechy** [reply]
Thanks @samarsault gave that a try on this branch: https://github.com/infor-design/enterprise-wc/pull/1289/files#diff-0f2ba0c2388feb41970f6db0773b0507712e9a1121575effdf4bbccec2ecedde

Seems its doing some looping https://github.com/infor-design/enterprise-wc/actions/runs/5047016563/jobs/9053293027#step:10:50372 

Did i get that all correct?

**Reply by samarsault** [reply]
The form e2e tests for which errors are shown pass fine locally. The range error is actually coming from ids-wizard. Just do `npm start` and go to `http://localhost:4300/ids-wizard/example.html`

<img width="1185" alt="Screenshot 2023-05-23 at 6 05 28 PM" src="https://github.com/percy/cli/assets/2987087/eeca76d5-0c6c-4fa5-9890-a5fdf03d0ae5">

Also can you try running the tests locally as well? (After the range fix)

**Reply by tmcconechy** [reply]
@samarsault i didnt run into this? Are you on branch `fix-percy-updates` and updated with `git pull` as i think we fixed this error a while ago.

![Screenshot 2023-05-23 at 10 26 14 AM](https://github.com/percy/cli/assets/814283/21fc7d1b-d802-4203-9690-3778873f453d)

**Reply by samarsault** [reply]
Yes I ran it on the latest `fix-percy-updates` branch. The CI should also be getting this. Can you also try running the tests locally once with the suggested changes and the range error fixed?
